### PR TITLE
Support for stable (unchanging) User Ids

### PR DESCRIPTION
**Pull request by esenciadev**
This adds support for using a Raven-generated ID for Users, so that changing a Users email does not affect their Id. The default behavior should be unchanged from the existing Identity package (other than a few bug fixes mentioned below), so existing users of the package will not have anything break. By passing in a configuration option to RavenDb.Identity `.AddRavenDbIdentityStores<AppUser>(o => o.StableUserId = true)` the Stable User Id mode is enabled. With this enabled when a user changes their email address, the compare/exchange value for them is recreated with the new Email, and the Email property on their document is updated.

Also in this PR are some minor bug fixes for things I found while implementing this:
1. `FindUserByEmail` now uses the compare/exchange value to get their actual UserId, then calls `Load()` with that Id. This resolves an issue where a user could change their email address, and another call very shortly after would get a stale result because it was querying an index.
2. `Update` had error handling code that would attempt to recreate the compare/exchange value if an update failed, but the key value it was passing to the methods had already been formatted to the compare/exchange key, which resulted in the key being double-formatted (eg: `emails/emails/foo@bar.com`)
3. A few places that could pass a CancellationToken but weren't.

@JudahGabriel - I know you prefer `.ToLowerInvariant()` and I agree with you on that preference, but the dotnet core Identity components which are calling into this `UserStore`, such as `UserManager` are [expecting email/username to use ToUpperInvariant()](https://github.com/dotnet/aspnetcore/blob/master/src/Identity/Extensions.Core/src/UpperInvariantLookupNormalizer.cs). Given that, maybe Raven Identity should use the same approach for normalizing email & username? This `UserStore` could take a dependency on `ILookupNormalizer` from the framework and then have consistent behavior for however the user wants to configure normalization. If you are OK with this approach, I can make that change in a separate PR.

### CODE REVIEW

File: RavenDB.Identity/RavenIdentityOptions.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    public class RavenIdentityOptions
+    {
+        /// <summary>
+        /// If set to true, this will use a server generated id for the User Id. In this case, changing the user email address
+        /// has no effect on the User Id. If set to false, changing the users email address will change their User Id. Default
+        /// is false.
+        /// </summary>
+        /// <remarks>It is critical this value is not changed once users have been stored in RavenDb. Doing so will result
+        /// in strange behavior and probably loss of data.</remarks>
+        public bool StableUserId { get; set; } = false;
```
Comment (JudahGabriel) [inline RavenDB.Identity/RavenIdentityOptions.cs:15]: Can you change this to an enum instead? And update the name correspondingly? I feel "stable user ID" is vague here.

File: Readme.md [hunk 2]
```diff
@@ -58,6 +58,21 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 }
 ```
 
+3.a. If you want to be able to change the email address of a user without also changing their user id, enable `StableUserId` mode:
```
Comment (JudahGabriel) [inline Readme.md:61]: This should be an addition to the end of the readme (e.g. Modifying RavenDB Conventions), not in the getting started section. Getting started, users can use the defaults.

**Comment by JudahGabriel** [comment]
@esenciadev thanks for this - I added some comments above, can you have a look?

File: RavenDB.Identity/UserStore.cs [hunk 3]
```diff
@@ -525,7 +525,7 @@ public Task SetSecurityStampAsync(TUser user, string stamp, CancellationToken ca
         public Task SetEmailAsync(TUser user, string email, CancellationToken cancellationToken)
         {
             ThrowIfDisposedOrCancelled(cancellationToken);
-            user.Email = email ?? throw new ArgumentNullException(nameof(email));
+            user.Email = email.ToLowerInvariant() ?? throw new ArgumentNullException(nameof(email));
```
Comment (JudahGabriel) [inline RavenDB.Identity/UserStore.cs:624]: This should be email?.ToLowerInvariant()

**Comment by cadilhac** [comment]
Thanks @JudahGabriel for writing this library and thanks to @esenciadev for this PR because I can not conceive this library with an email as the ID. Users need to change their email without changing references everywhere in the DB and I must be able to easily load a user thanks to its integer ID (from a /user/33 web page for instance).

Question: what prevents it from already being merged? Not final yet?

**Comment by esenciadev** [comment]
@cadilhac It's my bad, I got distracted with other stuff and failed to make a few updates that needed to happen. I'll get those changes in this week

**Comment by cadilhac** [comment]
> My changes to this do not result in a pure integer ID, it will create a hi/low ID like "users/123-A" and "users/543334-B". If you need a pure integer ID the best solution is to populate the ID property of your user entity before you add it to the store with a value that ends in the pipe character, like 'users|'. This will cause Raven to generate a cluster unique integer ID. Note that there are performance costs associated with using this technique.

Don't worry for that. I already have a solution in place with a custom AsyncMultiDatabaseHiLoIdGenerator set in the conventions. Being on a single server, I don't need that tag at the end of the id.

File: RavenDB.Identity/RavenIdentityOptions.cs (hunk 1 above)
Comment (ellipsis-apps) [reply RavenDB.Identity/RavenIdentityOptions.cs:15]: How about "ImmutableUserId".

**Comment by JudahGabriel** [comment]
I'm going to merge this and do a few tweaks based on the above comments.

**Comment by JudahGabriel** [comment]
Looking at this PR closer, I think it conflates 2 things:

- User uniqueness (email based)
- User ID generation

Those things can be different; it's entirely reasonable to enforce uniqueness by email. However, we may not want user IDs to be based on email address because users can change their email address.

I'm updating the code accordingly.

**Comment by JudahGabriel** [comment]
Done. I've deployed Raven.Identity v8, which supports configuring how user IDs are generated.

```csharp
// Use server-generated IDs, e.g. "AppUsers/0001-A"
services
    .AddRavenDbIdentityStores<AppUser>(o => o.UserIdType = UserIdType.ServerGenerated);

// Or, use username-based IDs, e.g. "AppUsers/johndoe"
services
    .AddRavenDbIdentityStores<AppUser>(o => o.UserIdType = UserIdType.UserName);
```

Read more about it [here](https://github.com/JudahGabriel/RavenDB.Identity/#changing-how-user-ids-are-generated).

**Comment by JudahGabriel** [comment]
Just a quick update to this: in Raven.Identity v8.0.2, we're Id-agnostic. Meaning, we don't care what ID you use. If you don't supply an ID when creating the user, Raven will assign one based on its own rules.

```c#
// ID will be "AppUsers/1-a"; Raven's default.
var user = new AppUser(...) { Id = null };
userManager.CreateAsync(user, ...);

// ID will be incrementing, e.g. "AppUsers/42"
var user = new AppUser(...) { Id = "AppUsers|" }; // cluster-wide increment
userManager.CreateAsync(user, ...);

// ID will be server-generated, e.g. "AppUsers/0000001-A"
var user = new AppUser(...) { Id = "AppUsers/" };
userManager.CreateAsync(user);

// ID will be whatever we set it to
var user = new AppUser(...) { Id = "AppUsers/johndoe@email.com" };
userManager.CreateAsync(user);
```

In short, Raven.Identity doesn't care what your ID is. We use email for duplicate checking, but that is ID-agnostic. This leaves ID generation left to your app, and if your app doesn't care, ID generation is left to Raven itself.

I think this is the best of both worlds. 😎

**Pull request merged**